### DISCUSSION TITLE
Using GLLVM to study fire salamander colouration

### DISCUSSION THREAD

**Discussion by Max-Muehlenhaupt** [JenniNiku/gllvm#141] [Q&A]
Hi all,
I would like to use gllvm to analyze a dataset on fire salamander dorsal patterns. I have data on 15 individuals per sex from 20 different locations (= population). Overall, I have 598 data points (one population has only 13 females and 15 males). Each fire salamander was assigned a specific colour pattern out of five possible patterns. Thus, per fire salamander I have five response variables (for the five colour patterns) and each fire salamander has a "1" for the colour pattern it was assigned to and four "0"s for the four colour patterns it was not assigned to. 
Now, I want to find out if males are more likely to have a specific colour pattern than females while controlling for the effect of population (fire salamanders from the same population are more likely to have the same dorsal colour pattern). 
I first tried this model:
fit_env_bin1 <- gllvm(Y, X, formula = ~ Sex + Population,
                      family = "binomial", num.lv = 1)
Which worked fine, but here, Population is coded as a fixed effect and I would rather have it as a random effect to control for the non-independence of samples from the same population. I tried using 
randomX and row.eff but it didn't work out, yet. 
Any ideas how I could code population as a random effect in a gllvm? 
I couldn't really find an answer from [this post](https://github.com/JenniNiku/gllvm/issues/16).

Thanks in advance!

**Comment by BertvanderVeen** [comment]
Hello Max! Sorry I didn't get back to you before. If "population" is in X, you should be able to do this:

fit_env_bin1 <- gllvm(Y, X, formula = ~ Sex,family = "binomial", num.lv = 1, row.eff = ~(1|Population)).

Let me know if that doesn't work for some reason and we will figure something out.

**Reply by Max-Muehlenhaupt** [reply]
Hi Bert,

thanks for your help. When running this code, I get this warning message:

1: In xtfrm.data.frame(x) : cannot xtfrm data frames
2: In gllvm.TMB(y, X = X, lv.X = lv.X, formula = formula, lv.formula = lv.formula,  :
  Error in optim(objr$par, objr$fn, objr$gr, method = "BFGS", control = list(reltol = reltol,  : 
  non-finite value for optim
3: In gllvm(Y, X, formula = ~Sex, family = "binomial", num.lv = 1,  :
  Algorithm converged to infinity, try other starting values or different method.

Y is a matrix of 5 columnes (one per colour pattern) and 598 rows (one per individual9 where each individual has a single "1" and four "0"s according to its colour pattern on the back.
X is a matrix of 2 columns (sex and population) and 598 rows (one per individual).

Do you know what the issue might be?

**Reply by BertvanderVeen** [reply]
Which version of gllvm are you using? Try including `starting.val = "zero"` in your model code, does that change anything?

**Reply by Max-Muehlenhaupt** [reply]
When adding the code (fit_env_bin1 <- gllvm(Y, X, formula = ~ Sex,family = "binomial", num.lv = 1,
                      row.eff = ~(1|Population), starting.val = "zero")), I get:

Algorithm converged to infinity, try other starting values or different method. 
Warning message:
1: In gllvm.TMB(y, X = X, lv.X = lv.X, formula = formula, lv.formula = lv.formula,  :
  Error in optim(objr$par, objr$fn, objr$gr, method = "BFGS", control = list(reltol = reltol,  : 
  non-finite values for optim 

2: In gllvm(Y, X, formula = ~Sex, family = "binomial", num.lv = 1,  :
  Algorithm converged to infinity, try other starting values or different method.
Previously, I have been using gllvm version 1.4.1 but I updated to 1.4.3 and the warning message keeps showing up.

**Reply by BertvanderVeen** [reply]
As a final attempt you can try to set `method = "LA"` although that is unlikely to make a difference, I think.

It's hard to say what's going on without seeing the data, but the fact that you only have 5 response variables and is likely to cause issues in estimating the variance of the random effect, though 20 populations should be sufficient... How many non-zero observations do you have per population?

**Reply by Max-Muehlenhaupt** [reply]
Hey Bert,
I see. I've tried using `method = "LA` but the same warning appears. 
So, basically each row corresponds with one individual with a specific sex and population-identity (20 populations, 15 males and 15 females per population, so overall 30 individuals per population). As my dependent variables I have 5 variables, one per colour pattern, and each individual has **only one 1** for the specific colour pattern it was assigned to and **four 0s** for the colour patterns it wasn't classified as.
I could also restructure the data set (e.g., grouped by population x sex and thus, compositional) but I thought the way, I have the data structrued at the moment, makes it easier to control for the effect of population on the chances of being a specific colour pattern (i.e., our random effect). We already know that population is an important factor in the chances of having a specific colour pattern for a fire salamander (based on ANOSIM results) but I'm more interested in the effect of sex.

**Reply by BertvanderVeen** [reply]
Thanks, that was helpful.

OK, so 30 individuals per population and 20 populations sounds OK. I think the problem might be the 5 response variables; 1 LV for 5 responses if a bit overkill, since with 5 responses dimension reduction is hardly necessary, especially for a binary response with one observation per row. So, if you can think of a smart way to restructure the data, that might be the way to go.
